File: 0773. Sliding Puzzle/Solution0773.py
# modify solution to general size puzzle board
# lintcode https://www.lintcode.com/problem/794/
# leetcode https://leetcode.com/problems/sliding-puzzle/description/
class Solution0773:
    def slidingPuzzle(self, board: List[List[int]]) -> int:
        src = self.matrix_to_string(board)
        tgt = "123450"

        from collections import deque
        queue = deque([src])
        distance = {src: 0}

        while queue:
            curt = queue.popleft()
            if curt == tgt:
                return distance[tgt]
            for next in self.get_next(curt, board):
                if next in distance:
                    continue
                queue.append(next)
                distance[next] = distance[curt] + 1
        return -1
            
    def get_next(self, curt: string, board: List[List[int]]):
        states = []
        dirs = ((0,1),(1,0),(0,-1),(-1,0))
        zero_idx = curt.find('0')
        x, y = zero_idx // len(board[0]), zero_idx % len(board[0])
        for i in range(4):
            x_, y_ = x + dirs[i][0], y + dirs[i][1]
            if 0 <= x_ < len(board) and 0 <= y_ < len(board[0]):
                next_state = list(curt)
                next_state[x*len(board[0])+y] = next_state[x_*len(board[0])+y_]
                next_state[x_*len(board[0])+y_] = '0'
                states.append("".join(next_state))
        return states

    def matrix_to_string(self, board: List[List[int]]) -> str:
        str_list = []
        for i in range(len(board)):
            for j in range(len(board[0])):
                str_list.append(str(board[i][j]))
        return "".join(str_list)

File: 0773. Sliding Puzzle/Solution0773.java
// # modify solution to general size puzzle board
// # lintcode https://www.lintcode.com/problem/794/
// # leetcode https://leetcode.com/problems/sliding-puzzle/description/
public class Solution0773 {
    public int slidingPuzzle(int[][] board) {
        int row = board.length, col = board[0].length;
        String src = print(board), tgt = "123450";
        Deque<String> queue = new ArrayDeque<>();
        Set<String> visited =  new HashSet<>();
        Map<String, Integer> dist = new HashMap<>();
        queue.offerLast(src);
        dist.put(src, 0);

        while (!queue.isEmpty()) {
            String curt = queue.pollFirst();
            if (curt.equals(tgt)) {
                return dist.get(tgt);
            }
            for (String next : getNext(curt, board)) {
                if (dist.containsKey(next)) {
                    continue;
                }
                dist.put(next, dist.get(curt) + 1);
                queue.offerLast(next);
            }
        }
        return -1;
    }

    public static final int[][] DIRS = new int[][]{{0,1},{1,0},{0,-1},{-1,0}};

    private List<String> getNext(String curt, int[][] board) {
        List<String> nexts = new ArrayList<>();
        int row = board.length, col = board[0].length;
        int zeroIdx = curt.indexOf('0');
        int r = zeroIdx/col, c = zeroIdx%col;
        for (int[] dir : DIRS) {
            int r_ = r + dir[0], c_ = c + dir[1];
            if (r_ < 0 || r_ >= row || c_ < 0 || c_>= col) {
                continue;
            }
            String next = swap(curt, r*col+c, r_*col+c_);
            nexts.add(next);
        }
        return nexts;
    }

    private String print(int[][] board) {
        int row = board.length, col = board[0].length;
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < row; i++) {
            for (int j = 0; j < col; j++) {
                sb.append((char) (board[i][j] + '0'));
            }
        }
        return sb.toString();
    }

    private String swap(String str, int i, int j) {
        StringBuilder sb = new StringBuilder(str);
        sb.setCharAt(i, str.charAt(j));
        sb.setCharAt(j, str.charAt(i));
        return sb.toString();
    }
}